### ISSUE TITLE
Add support for Canon CR3 raw files

### ISSUE THREAD

**Issue by konig12** [darktable-org/darktable#2170]
Canon CR3 raw files are currently not supported. These are used by a few of Canon's new cameras at this point. At a minimum, the M50, SX70 HS, EOS R, and EOS RP. https://raw.pixls.us/ has samples for the M50, EOS R, and EOS RP.

Someone appears to have already done a lot of the reverse engineering here:
https://github.com/lclevy/canon_cr3

Let me know if there is anything that can be done to help, including testing, more samples, or development assistance.

**Comment by DawidLoubser** [comment]
I just wanted to add that I have also uploaded sampler RAW and C-RAW from an EOS R - and I would do anything else in my power to help support the efforts to reverse-engineer the CR3 format and get it into Darktable.

**Comment by arigit** [comment]
On the ongoing effort to reverse-engineer the proprietary CR3 format: https://github.com/lclevy/canon_cr3

**Comment by tim-hilt** [comment]
cr3-support is now available in LibRaws master-branch.

**Comment by DawidLoubser** [comment]
That is fantastic news!! Well done to all that helped!

**Comment by utdemir** [comment]
Nope, still not solved in master.

**Comment by carloszacu** [comment]
I have a Canon EOS Rebel SL3 (also known as EOS 250D and EOS Kiss X10) and I still can´t get my files read by DarkTable, I had to use the Adobe DNG Converter to make the files usable with the app but still no direct support, highly interested in this compatibility, if you need any sample shots or something I would love to help.

**Comment by JackSwett** [comment]
For those needing a workaround, you can use Adobe's DNG converter [https://helpx.adobe.com/photoshop/using/adobe-dng-converter.html](url)

Obviously native support would be very much appreciated!

**Comment by RReverser** [comment]
Yup, also using DNG converter, but, particularly with C-RAW files, having to decompress them just for post-processing really bloats the disk.

**Comment by isolin** [comment]
There is no Linux version of DNG Converter. I have to take RAW+JPEG and hope that one day I will be able to process the RAWs in Darktable (being my favourite tool). At least RawTherapee added limited support recently. It's a shame Adobe ignores Linux and I seriously blame them for not warning their customers :(

**Comment by RReverser** [comment]
FWIW there is a reverse-engineering spec here https://github.com/lclevy/canon_cr3 as well as the mentioned earlier support in other libraries like LibRAW, but, my understanding is that devs are mainly concerned about legality of using a reverse-engineered spec?

I wonder how LibRAW deals with this issue? (if it does)

**Comment by johnny-bit** [comment]
We know :) Pretty much CR3 is possible, just needs exiv2 support and it's stalled on exiv2 supporting it.

**Comment by 1div0** [comment]
Almost one year already. The speed of Open Sauce development is incredible.

**Comment by johnny-bit** [comment]
@1div0:
>Almost one year already. The speed of Open Sauce development is incredible.

Has Canon released non-NDA specs for CR3? Has Canon cleared open source of any possible patent infrigment? Had nobody objected to CR3 support being done in exiv2? Have you contacted Canon and asked them to provide (or at least open line of communication) specs?

Anyway, slowdown on exiv2 side is mostly caused by exiv2 community-member(s) objecting to including possibly patent infringing code (it wouldn't be, I asked law advisor) and thus halting development and discouraging it. I thought I could work on it, but life & paying bills got in the way so I hope Mr. Mills or somebody else will be able to implement relevant stuff on exiv2 side. On darktable's side support is possible as it currently resides on experimental branch (no, that can't be enabled without exiv2 working)

**Comment by RReverser** [comment]
I wonder if, as a workaround, Darktable could provide LibRAW support in form of a plugin? I know it used to use LibRAW in the past, and then switched to custom decoder for performance reasons (I think?), but could LibRAW at least be used as a fallback for unsupported formats so that it would solve cases like this?

**Comment by mathbagu** [comment]
I'm also interested in the support of CR3 in darktable. Can anyone give us a status? How can we help?

**Comment by SebastianPfliegel** [comment]
> There is no Linux version of DNG Converter 👎 I have to take RAW+JPEG and stick with JPEGs for now, hoping that one day I will be able to process the RAWs in Darktable (being my favourite tool). At least RawTherapee added limited support recently. It's a shame Adobe ignores Linux, it's a shame they don't opensource the CR3 specification and I seriously blame them for not warning their customers :(

@isolin the Adobe converter works flawless using Wine. Not the best solution, but works for me in the meantime.

**Comment by daniel-lucio** [comment]
For those impatient:
https://inside-out.xyz/technology/editing-cr3-canon-raw-format-with-darktable-on-linux.html

Note that I save in RAW not cRAW.

**Comment by RReverser** [comment]
Note that solution from @daniel-lucio is also converting to DNG, like couple others in the thread. It's not a support in Darktable.

**Comment by daniel-lucio** [comment]
> Note that solution from @daniel-lucio is also converting to DNG, like couple others in the thread. It's not a support in Darktable.

I know, that's why I said for the impatient.

**Comment by aurelienpierre** [comment]
Closing here, there is a similar feature request open on Rawspeed project, and Exiv2 is the sole responsible for that lack of support.

**Comment by RReverser** [comment]
For future people coming across this thread and trying to find a RawSpeed issue mentioned in the last comment: https://github.com/darktable-org/rawspeed/issues/121

@aurelienpierre That said, can you respond to the suggestion above (in https://github.com/darktable-org/darktable/issues/2170#issuecomment-657280851)? That one is specifically not related to RawSpeed, but only to Darktable.

**Comment by LebedevRI** [comment]
> I wonder if, as a workaround, Darktable could provide LibRAW support in form of a plugin? I know it used to use LibRAW in the past, and then switched to custom decoder for performance reasons (I think?), but could LibRAW at least be used as a fallback for unsupported formats so that it would solve cases like this until a better version is available?

FWIW the current darktable development trend seems to be on track to undo/change
most of the previous decisions, so that doesn't seem outlandish to me.

Other than that, all the same reasons that led to the current camera support landscape
are still there; undoing that will lead to large negative consequences.

**Comment by RReverser** [comment]
> undoing that will lead to large negative consequences

I wouldn't necessarily "undo" it, but merely provide LibRAW (or maybe even general decoder registration API for plugins) as an alternative / fallback.

**Comment by parafin** [comment]
No one is working on it and I never saw anyone mentioning he wants to do it. Hopes that answers this question.

**Comment by 1div0** [comment]
Talk is cheap.

**Comment by RReverser** [comment]
> No one is working on it and I never saw anyone mentioning he wants to do it. Hopes that answers this question.

The question wasn't whether someone is working on it, so not really, it doesn't.

The question was whether it's a suitable path to supporting new formats that was or would be considered and accepted by the core team, or it's a no-go and the previous removal of LibRAW is "set in stone".

**Comment by parafin** [comment]
Personally I don't see the point, CR3 support is moving forward with current software stack, and it was the only problem so far if I remember correctly. So benefits will basically be nonexistent for a large amount of work and testing. I'm not sure if such PR would be accepted (I wouldn't).

**Comment by RReverser** [comment]
> Personally I don't see the point, CR3 support is moving forward with current software stack

I see. My understanding was that it's blocked on legal issues / questions, which won't be resolved anytime soon, so "delegating" to a 3rd-party library that user would need to install separately seemed like an easier path forward.

**Comment by johnny-bit** [comment]
Legal issues/questions have been addressed and process for developing necessary parts is ongoing now. @1div0 is working on heif/heic support in exiv2 and as far as I'm aware exiv2 is now willing to accept patches for other ISOBMFF based files. If time allows I probably try to do something myself so :)

**Comment by parafin** [comment]
See https://github.com/Exiv2/exiv2/issues/1229#issuecomment-727190658

**Comment by parafin** [comment]
So yeah, time is better spent going in this direction instead of (re-)integrating libraw.

**Comment by RReverser** [comment]
@johnny-bit Oh, that's great news - I've only seen the earlier mentioned issue on the RawSpeed repo, where the last comment / update didn't sound as hopeful. https://github.com/darktable-org/rawspeed/issues/121#issuecomment-703223293

Thanks!

**Comment by aurelienpierre** [comment]
Adding LibRAW support in darktable is not the plan. Anyway we rely on Exiv2 in other parts of the soft, even if squeezing-in LibRAW to circumvent Rawspeed limitation, it will not make lens detection possible anyway so it will be at best a nasty workaround.

**Comment by luisalves00** [comment]
Not sure how helpful it is, but I found this: [Canon CR3 support in Linux using ART, RawTherapee fork](https://garridodiaz.com/canon-cr3-support-in-linux-using-art-rawtherapee-clone). At least now I can work with the CR3 files.

**Comment by MStraeten** [comment]
lacking CR3 support is not a technical thing but related to an ambiguous legal situation in exiv2 library... 
the exiv2 team is a bit more careful in this respect

**Comment by luisalves00** [comment]
@MStraeten great...hopefully this is solved quickly. Don't understand why Canon don't release the lib to read their format as opensource. Next time I'll buy another brand.

**Comment by MStraeten** [comment]
They don’t need to release a lib, just confirm it’s no copyright infringement to decode cr3 for reading...

**Comment by goeland86** [comment]
@MStraeten the exiv2 team has not touched the topic since ~ June 2020 at the time I'm posting this.

Is there anything end-users of DT like me can try to do to unblock the exiv2 situation? I'm starting to get really frustrated, because my Shotwell/DT workflow was working really well... Until I changed cameras to a 90D, and I'm more than a little concerned that exiv2 refuses to support the format.

**Comment by MStraeten** [comment]
the discussion on cr3 is better at the proper place - https://github.com/Exiv2/exiv2/issues/1229#

**Comment by alexvanderberkel** [comment]
There will be a meeting on Saturday to discuss how this issue can be resolved. The idea is to form a task force from different projects to resolve the missing part for CR3. Any help is welcome. 

Topic: Support for ISOBMFF Files (AVIF, HEIF, CR3)
Time: Jan 9, 2021 07:30 PM Amsterdam, Berlin, Rome, Stockholm, Vienna

Join Zoom Meeting
https://us05web.zoom.us/j/82168885673?pwd=K3pqRThERVNQK2JQR2ZweStwS2Nkdz09

Meeting ID: 821 6888 5673
Passcode: L1B3Hz


Furhter discussion are ongoing here: https://app.element.io/#/room/#exiv2-chat:matrix.org

**Comment by Jurze** [comment]
Thanks to everyone whos helping. Are there any news?

**Comment by DerfOh** [comment]
> There will be a meeting on Saturday to discuss how this issue can be resolved. The idea is to form a task force from different projects to resolve the missing part for CR3. Any help is welcome.
> 
> Topic: Support for ISOBMFF Files (AVIF, HEIF, CR3)
> Time: Jan 9, 2021 07:30 PM Amsterdam, Berlin, Rome, Stockholm, Vienna
> 
> Join Zoom Meeting
> https://us05web.zoom.us/j/82168885673?pwd=K3pqRThERVNQK2JQR2ZweStwS2Nkdz09
> 
> Meeting ID: 821 6888 5673
> Passcode: L1B3Hz
> 
> Furhter discussion are ongoing here: https://app.element.io/#/room/#exiv2-chat:matrix.org

How'd it go? I too am still waiting for this feature.

**Comment by johnny-bit** [comment]
Be patient, development is ongoing 😃

czw., 28 sty 2021, 19:10 użytkownik Fredrick P. <notifications@github.com>
napisał:

**Comment by Jurze** [comment]
That's nice to hear. I appreciate your work just because i never could do this on my own. Take your time. Thank you.

**Comment by quovadit** [comment]
@johnny-bit : is there anything I can do to help?

I would like to buy the canon r5, but of course it does not make any sense unless dt supports cr3

**Comment by johnny-bit** [comment]
@quovadit - the https://github.com/darktable-org/darktable/pull/8340 PR by @1div0 does enable ISOBMFF reading support coming in eviv2 v 0.27.4, which is prerequisite for CR3 support.

This is the last step needed before rawspeed work, so after that PR is merged the missing puzzle would be in rawspeed. If you wish to help then I guess best bet would be to immerse yourself in rawspeed and coordinate with @LebedevRI

**Comment by mfg92** [comment]
I just want to inform you that there is now a RC for EXIV2 with support for bmff/cr3 files:
https://github.com/Exiv2/exiv2/releases/tag/v0.27.4-RC2

**Comment by axettone** [comment]
I have a Canon R6, how can I contribute?

**Comment by johnny-bit** [comment]
posting samples on raw.pixls.us is usually the first step ;) next is actual CR3 work as descibedy in my commend https://github.com/darktable-org/darktable/issues/2170#issuecomment-787050763 :)

**Comment by axettone** [comment]
@johnny-bit thank you for your reply. As far as I can see there are already two .CR3 (RAW and C-RAW) files on raw.picls.us. I don't understand if it's useful tu submit other photos.

**Comment by mirecta** [comment]
How can i help ? Exiv2 i ready so now , so what is next step ?

**Comment by cytrinox** [comment]
> How can i help ? Exiv2 i ready so now , so what is next step ?

The last todo is https://github.com/darktable-org/rawspeed/issues/121, implementing a cr3 decoder in rawspeed to read the raw pixels.

**Comment by alexvanderberkel** [comment]
@axettone  I uploaded 120 images today in all possible ISO steps from  Canon EOS R in the format 1:1 to raw.pixls.us. 60 for RAW+CRAW with dual pixel support on and 60 images for RAW + CRAW with dual pixel support off. Put CC0 license in the meta data as well and I shared them as well with @cytrinox via wetransfer etc. 

I put the camera on the tripod and adjusted the settings accordingly so that the image was well exposed in the histogram. 

Could you do the same for the R6 and anybody else for the M50 Mk2, RP or 1DX Mk3?

**Comment by konig12** [comment]
@alexvanderberkel I am happy to provide any needed images from the EOS RP, but I would like to clarify what exactly is needed first. raw.pixls.us specifically states that images from different ISO levels are not needed. Is that still accurate? If you point me to the updated requirements I'm happy to capture and upload the needed images.

**Comment by cytrinox** [comment]
I need all ISO levels (please enable 1/3 steps for ISO 125, 160 ...) as well as ISO L (50) and HI-ISO Levels (51.200, ...).
Please rename the file with pattern Canon_EOS_RP_RAW_ISO_100_no_dualpixel.CR3 or Canon_EOS_RP_CRAW_ISO_100_no_dualpixel.CR3.
For Canon, all ISO levels are required for Black/White level adjustments. You can share the packed ZIP file directly with me, use Dropbox, WeTransfer or anything else. I will make the whole CR3 sample package available to raw.pixls.us after some file name cleanup.

**Comment by montagdude** [comment]
@cytrinox Similar to the question about, I can provide files for the EOS M50 if that would help. Let me know. I'd love to be able to use darktable for this camera so I could ditch DPP4+Wine.

**Comment by johnny-bit** [comment]
@montagdude - check https://discuss.pixls.us/t/call-for-extended-cr3-samples-all-iso-combinations-needed/24959?u=johnny-bit for what's needed more :)

Currently we're loking for M50 mk2 and G7 X mk3. For the rest of cr3 generators we have full set :)

**Comment by montagdude** [comment]
@johnny-bit I've got the M50 mk1, so I guess I can't help fill in what's needed. Bummer.

**Comment by dov** [comment]
Where can I find instructions of how to build darktable with CR3 support?

Here is what i tried:

1. Checked out darktable HEAD (master) from git
2. Did `git submodule init; git submodule update`
3. cd'd into `src/external/darktable` and did 'git config remote.origin.url https://github.com/cytrinox/rawspeed.git` and `git fetch`. This corresponds to the currently opened rawspeed CR3 pull request.
4. Changed the darktable `.gitmodules` file and updated the `rawspeed` section to:
```
[submodule "src/external/rawspeed"]
	path = src/external/rawspeed
        url = https://github.com/cytrinox/rawspeed.git
	branch = canon_cr3
```
5. Did `git submodule update --remote`
5. Go to repo top directory. Then ` mkdir build; cd build; cmake .. ; make ; make install`

Everything compiled smoothly on my Fedora 34 box (after adding some additional packages). But when trying to import .CR3 files (from my Canon R6) they did not appear in the file list. What did I miss? Is it necessary to enable CR3 files somewhere?

Thanks!

**Comment by MStraeten** [comment]
You need also to install exiv2 0.27.4 RC3 compiled with bmff support (not enabled by default)

**Comment by dov** [comment]
Thanks for the tip. I tried with exiv2 HEAD from github. The exiv2.pc files sais it is version 1.0.0.9. bmff is enabled by default. All compiled and installed and I have verified with ldd that I'm using the new exiv2. But still, the CR3 files are not recognized. What am I missing?

**Comment by MStraeten** [comment]
search the output of cmake for
`-- Exiv2 >= 0.27.4 found with ISOBMFF support (CR3, AVIF, HEIF)`
or 
```
-- Found Exiv2: /usr/local/lib/libexiv2.dylib (Required is at least version "0.24")
-- Looking for EXV_ENABLE_BMFF
-- Looking for EXV_ENABLE_BMFF - not found
-- Exiv2 >= 0.27.4 found, but was not compiled with ISOBMFF support (CR3, AVIF, HEIF)
```
if it's not found then it might be similar to #9024 on osx 

in this case you can override the check for bmff support by replacing 
`check_cxx_symbol_exists(EXV_ENABLE_BMFF "exiv2/exiv2.hpp" HAVE_EXV_ENABLE_BMFF) ` by `set(HAVE_EXV_ENABLE_BMFF 1)`
in darktable/src/CMakeLists.txt

**Comment by dov** [comment]
@MStraeten Thanks! It worked!

**Comment by montagdude** [comment]
Thanks from me too, it's working so far.

**Comment by mfg92** [comment]
Congratulations on the great release of darktable 3.6!
Are there any chances that we will see CR3 support added in 3.6.1 or 3.6.x?

**Comment by mega-stoffel** [comment]
> search the output of cmake for
> `-- Exiv2 >= 0.27.4 found with ISOBMFF support (CR3, AVIF, HEIF)`
I indeed have this output when running "cmake build" in the darktable directory. And I compiled this exiv2 in the 1.0.0.9 version including this BMFF support. This line is in the output of the cmake build in the exiv2 directory:
> -- Building BMFF support:              YES
But it's I still can't load CR3 files. The file names are shown in the import dialogue and after importing even in the "Leuchttisch" view. But the preview shows just the skull and when I want to open it in the "Dunkelkammer", it gives me an error message, that it can't load the CR3 file.
I'm afraid I have to wait a bit longer for a newer release, which is capable of opening CR3 files (or a howto documentation for non-Linux-software-engineers).

**Comment by piratenpanda** [comment]
Are you sure you replaced rawspeed with cytrinox's branch @mega-stoffel? https://github.com/cytrinox/rawspeed/tree/canon_cr3

**Comment by mega-stoffel** [comment]
@piratenpanda - I'm afraid I haven't done that. I'll try it out. Thanks a lot for the hint! :-)

**Comment by montagdude** [comment]
After building the CR3 branch as described above, I can load and work with CR3 files from my EOS M50 in Darktable, but not samples from the G5 X mark II, a camera I'm considering buying. Any recommendations for that? I ran the files through exiv2 and didn't see any problems.

**Comment by piratenpanda** [comment]
> Any recommendations for that?

It's not yet supported by rawspeed

**Comment by montagdude** [comment]
I'm having a weird issue with M50 CR3 files with a new lens, the Sigma 30mm f/1.4 DC DN. The lens is recognized by exiv2, but when an image is imported into Darktable, it thinks it is the Sigma 56mm f/1.4 DC DN lens instead of the 30 mm. On export, the JPG metadata has the right lens. Running `exiv2 -pt IMG_2764.CR3` I get:

Exif.Photo.LensModel                         Ascii      35  30mm F1.4 DC DN | Contemporary 016

So I don't understand why Darktable reads this as the 56mm lens. Also, if I use the Adobe DNG converter and then import the DNG instead of the CR3, the lens is correct. The main trouble with having it identify the lens incorrectly is that it makes it tedious to apply lens corrections. Should I upload a sample RAW file somewhere?

**Comment by piratenpanda** [comment]
You can always do https://dev.exiv2.org/projects/exiv2/wiki/Lens_Recognition_in_Exiv2_v026_(and_later)/

and then in darktable, lighttable, "selected images" on the right side, "metadata" tab, "refresh exif data"

**Comment by montagdude** [comment]
Thanks for the quick response. My exif data doesn't seem to contain a LensIDNumber. Maybe that's the problem?

```
exiv2 -pv --grep lens/i IMG_2879.JPG 
0x0016 CanonCs      LensType                    Short       1  4208
0x0017 CanonCs      Lens                        Short       3  30 30 1
0x003d CanonFi      RFLensType                  SShort      1  0
0x0095 Canon        LensModel                   Ascii      74  30mm F1.4 DC DN | Contemporary 016
0xa432 Photo        LensSpecification           Rational    4  30/1 30/1 0/1 0/1
0xa434 Photo        LensModel                   Ascii      35  30mm F1.4 DC DN | Contemporary 016
0xa435 Photo        LensSerialNumber            Ascii      11  0000000000
```

**Comment by piratenpanda** [comment]
Would be 4208 for your lens

**Comment by montagdude** [comment]
Gotcha. Thanks, that works. I put this in ~/.exiv2:
`[canon]
4208=30mm F1.4 DC DN | Contemporary 016`

**Comment by piratenpanda** [comment]
> After building the CR3 branch as described above, I can load and work with CR3 files from my EOS M50 in Darktable, but not samples from the G5 X mark II, a camera I'm considering buying. Any recommendations for that? I ran the files through exiv2 and didn't see any problems.

https://github.com/darktable-org/rawspeed/pull/271#issuecomment-887004069 latest commit from @cytrinox added the remaining cameras

**Comment by stelmach-web** [comment]
Is there any easy way to install exiv2 or enable CR3 support for EOS R6 on Windows? (without any building/complex procedures)?

**Comment by MStraeten** [comment]
Yes, have a look at https://discuss.pixls.us/t/darktable-3-6-0-unofficial-build-with-cr3-support-osx-win10/25703

**Comment by LB4FH** [comment]
> Yes, have a look at https://discuss.pixls.us/t/darktable-3-6-0-unofficial-build-with-cr3-support-osx-win10/25703
> or https://discuss.pixls.us/t/current-win10-build-incl-cr3-support/25212/10 for latest development build

Great to see that people are spending time on implementing this. I tried both of those, 3.6.0 didn't work (seg fault), while 3.7.0+620_g74ad7ea3f is able to read CR3's off a M50 MK ii.

**Comment by drazde** [comment]
From the release notes of DT 3.6 
> Preliminary work toward full support for CR3 has begun. darktable's exif
> support can now read CR3 files if the proper exiv2 library version
> is found.

So for build DT with CR3 support is now necessary have the exiv2 uptodate (and build correctly); Is it correct?

On my system exiv2 is too old so I build from souce and install on my home. Now I don't know how to use it..

I try with 
```
export PATH="${HOME}/local/exiv2/bin:${PATH}"
export LDPATH="${HOME}/local/exiv2/lib64:${LDPATH}"

# Building test version and install in: $HOME/loca/darktable-test
./build.sh --prefix $HOME/local/darktable-test --build-type Release --install

```

But it still use the old Exiv2. How can I use the "local" version of Exiv2?

**Comment by MultiCoreNop** [comment]
I have built exiv2-0.27.4  after applying the attached patch.
[exiv-enable-bmff.patch.txt](https://github.com/darktable-org/darktable/files/6937204/exiv-enable-bmff.patch.txt)
Afterward darktable 3.6.0 with the attached patch.
[exiv-enable-bmff.patch.txt](https://github.com/darktable-org/darktable/files/6937216/exiv-enable-bmff.patch.txt)

To make sure, you're using the correct exiv-lib for the DT compile, LIBRARY_PATH / CPATH are your friends.
If everything 'lurks around' in a directory without being installed to /usr/local or something, use  LD_LIBRARY_PATH pointing to your exiv-version to start DT.
HTH

**Comment by michaelmayer2** [comment]
> From the release notes of DT 3.6
> 
> > Preliminary work toward full support for CR3 has begun. darktable's exif
> > support can now read CR3 files if the proper exiv2 library version
> > is found.
> 
> So for build DT with CR3 support is now necessary have the exiv2 uptodate (and build correctly); Is it correct?
> 
> On my system exiv2 is too old so I build from souce and install on my home. Now I don't know how to use it..
> 
> I try with
> 
> ```
> export PATH="${HOME}/local/exiv2/bin:${PATH}"
> export LDPATH="${HOME}/local/exiv2/lib64:${LDPATH}"
> 
> # Building test version and install in: $HOME/loca/darktable-test
> ./build.sh --prefix $HOME/local/darktable-test --build-type Release --install
> ```
> 
> But it still use the old Exiv2. How can I use the "local" version of Exiv2?

if your system uses pkg-config, you may want to set

```
export PKG_CONFIG_PATH="${HOME}/local/exiv2/lib64/pkgconfig:$PKG_CONFIG_PATH"
```

which will trigger the definition of LIBRARY_PATH, CPATH, ... for you.

**Comment by paolobenve** [comment]
I could compile darktable following @dov recipe (and subsequent advices), darktable works with CR3 files!

however, using an M200 sample image downloaded from internet and opening it I get:

`[colorin] `Canon EOS M200' color matrix not found!`

The same error is shown with images from other camera. I suppose it's something that went wrong on compilation.... but... what?

**Comment by piratenpanda** [comment]
You need this for the remaining cameras https://github.com/darktable-org/darktable/pull/9641

**Comment by EC1000** [comment]
I thought I had it worked out, since I had this line when compiling: 
```
-- Exiv2 >= 0.27.4 found with ISOBMFF support (CR3, AVIF, HEIF)
```

However, dt does not launch; here is the error I get when I tried to launch it:
```
/opt/darktable_master_cr3/bin/darktable: symbol lookup error: /opt/darktable_master_cr3/bin/../lib/darktable/libdarktable.so: undefined symbol: _ZN5Exiv210enableBMFFEb
```
I somebody has any idea, I would be gratefull. 
(I am very new to git and compiling, fell free to tell me if this question would be better on pixls.us for example)

**Comment by ClemGlacee** [comment]
If you're on Linux (Arch or any arch based distro) you could look into that: https://aur.archlinux.org/packages/darktable-cr3-git/
Worked great for me

**Comment by max-phase** [comment]
> ```
> /opt/darktable_master_cr3/bin/darktable: symbol lookup error: /opt/darktable_master_cr3/bin/../lib/darktable/libdarktable.so: undefined symbol: _ZN5Exiv210enableBMFFEb
> ```
> 
> I somebody has any idea, I would be gratefull.
> (I am very new to git and compiling, fell free to tell me if this question would be better on pixls.us for example)

Maybe you have another libexiv2.so on your system?
Try
```
LD_LIBRARAY_PATH=/path/to/your/isobmff-exiv-lib-dir /opt/darktable_master_cr3/bin/darktable
```

**Comment by EC1000** [comment]
That was it! Thanks a lot :+1: 
(if somebody else need to use this advice, it's `LIBRARY` not `LIBRARAY` obviously)

**Comment by max-phase** [comment]
If I could get a penny every time I make this typo...

**Comment by freebrowser1** [comment]
I have an EOS R6 and darktable 3.6.0 (macOS) but it still refuses to load CR3 files.
Is this library also available for macOS ?

**Comment by MStraeten** [comment]
Have a look at https://discuss.pixls.us/t/darktable-3-6-0-unofficial-build-with-cr3-support-osx-win10/25703 for 3.6 or https://discuss.pixls.us/t/current-osx-build/13213/349 for the latest development build

**Comment by piratenpanda** [comment]
3.6.0 does NOT support CR3 files yet. You can get unofficial builds with cr3 support here: https://discuss.pixls.us/t/darktable-3-6-0-unofficial-build-with-cr3-support-osx-win10/25703

**Comment by freebrowser1** [comment]
Thanks, that worked and DT can now read my .CR3 photos.

**Comment by groovenectar** [comment]
> Have a look at https://discuss.pixls.us/t/darktable-3-6-0-unofficial-build-with-cr3-support-osx-win10/25703 for 3.6 or https://discuss.pixls.us/t/current-osx-build/13213/349 for the latest development build

Seems that Linux has been left out these builds :( What would it take to include it? I check in here every couple weeks

**Comment by MStraeten** [comment]
> Seems that Linux has been left out these builds :( What would it take to include it? I check in here every couple weeks

quite simple: it needs somebody who has a linux system and finds cr3 support for linux even at an officially unstable stage to be useful enough to build the stuff (there seems to be some doing this) and provide the results to the community. Just checking doesn’t create binaries ;)

**Comment by groovenectar** [comment]
> > Seems that Linux has been left out these builds :( What would it take to include it? I check in here every couple weeks
> 
> quite simple: it needs somebody who has a linux system and finds cr3 support for linux even at an officially unstable stage to be useful enough to build the stuff (there seems to be some doing this) and provide the results to the community. Just checking doesn’t create binaries ;)

That's fine, but more specifically where are the files to do this?

**Comment by piratenpanda** [comment]
Which files are you talking about. Compilation guides have been posted and linked in this issue so that's a good start. Which distro are you running?

**Comment by goeland86** [comment]
The idea of binaries for any distro can be done with an appimage release
potentially... But that's just my opinion.

**Comment by groovenectar** [comment]
> Which files are you talking about? Compilation guides have been posted and linked in this issue so that's a good start. Which distro are you running?

Somehow from reading in here I thought the CR3 builds were some kind of special thing that's not in the repository yet. I'm running Void Linux.

**Comment by piratenpanda** [comment]
Yeah, it needs the canon_cr3 branch from @cytrinox found here: https://github.com/cytrinox/rawspeed/tree/canon_cr3. If Void linux has the latest exiv2 (0.27.4) release that's all you need to compile darktable on your own to gain cr3 support

**Comment by groovenectar** [comment]
> Yeah, it needs the canon_cr3 branch from @cytrinox found here: https://github.com/cytrinox/rawspeed/tree/canon_cr3. If Void linux has the latest exiv2 (0.27.4) release, with ISOBMFF support enabled that is, that's all you need to compile darktable on your own to gain cr3 support

Thank you!! Here it is @MStraeten. Checking in does help, and it's not in the official repo yet so one can't "just build it" without realizing where it is.

**Comment by piratenpanda** [comment]
To be fair, it has been mentioned quite often here already even with detailed guides. You might have missed that.

**Comment by groovenectar** [comment]
> To be fair, it has been mentioned quite often here already even with detailed guides. You might have missed that.

Yes I did miss that... And still don't see detailed guides after checking again though I see where people have been trying it with a slight bit of back and forth... Maybe it's not that straightforward, I can at least give it a try though

**Comment by piratenpanda** [comment]
Github hiding posts doesn't help either. This is I think the most straight forward guide when exiv2 is available with ISOBMFF support:
https://github.com/darktable-org/darktable/issues/2170#issuecomment-854006616

I use the following cmake command instead of cmake .. in step 6:
`cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/opt/darktable/ ..`

**Comment by groovenectar** [comment]
> Github hiding posts doesn't help either. This is I think the most straight forward guide when exiv2 is available with ISOBMFF support:
> [#2170 (comment)](https://github.com/darktable-org/darktable/issues/2170#issuecomment-854006616)
> 
> I use the following cmake command instead of cmake .. in step 6:
> `cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/opt/darktable/ ..`

This looks like the ticket indeed, thank you!!!!

**Comment by piratenpanda** [comment]
Make sure you follow https://github.com/darktable-org/darktable/issues/2170#issuecomment-854152173 as well to check if your exiv2 supports ISOBMFF

**Comment by groovenectar** [comment]
I ran into some issues following https://github.com/darktable-org/darktable/issues/2170#issuecomment-854006616...


Firstly, `src/external/darktable` didn't exist in step 3... I created it, but I don't that made a difference... Had some issues with updating the submodule as well... I ended trying a few things, including directly editing the `.git/config` file of the `rawspeed` submodule to change it to the CR3-enabled repo and branch... I did a `fetch --all` in the submodule directory and then just checked out the branch in the worktree....

Then, running `cmake` returns this:

```
CMake Error at /usr/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
  (Required is at least version "4.5")
Call Stack (most recent call first):
  /usr/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.21/Modules/FindOpenMP.cmake:544 (find_package_handle_standard_args)
  CMakeLists.txt:89 (find_package)
```

Something to do with how `cmake` is packaged in Void I'm presuming... I installed `extra-cmake-modules` as well as `cmake` for doing this...

**Comment by piratenpanda** [comment]
Yeah needs to be /external/rawspeed

regarding the error maybe this helps:
https://stackoverflow.com/questions/48825416/missing-openmp-c-flags-openmp-c-lib-names

**Comment by groovenectar** [comment]
> Yeah needs to be /external/rawspeed

Alright, that makes sense, that's where I was digging around with the configs.. though I guess the `cmake` error is a show-stopper for me sadly...

> regarding the error maybe this helps:
> https://stackoverflow.com/questions/48825416/missing-openmp-c-flags-openmp-c-lib-names

Thank you, I did see that though it looks to be OSX and I'm guessing that something will go wrong for me

**Comment by piratenpanda** [comment]
Do you have all openmp packages installed? Not only the devel packages that is

**Comment by groovenectar** [comment]
Here's what I get when searching for `openmp`:

```
$ xbps-query -Rs openmp
[*] libgomp-10.2.1pre1_3       GNU Compiler Collection - OpenMP v4.0 library
[-] libgomp-devel-10.2.1pre1_3 GNU Compiler Collection - OpenMP v4.0 library - development files
[-] libomp-12.0.0_3            Low Level Virtual Machine - Clang OpenMP support library
[-] libomp-devel-12.0.0_3      Low Level Virtual Machine - Clang OpenMP support library - development files
[-] libopenmpi-4.1.1_1         High Performance Message Passing Library - runtime libraries
[-] libopenmpt-0.5.10_1        Cross-platform C & C++ library to decode tracked music files (modules)
[-] libopenmpt-devel-0.5.10_1  Cross-platform C & C++ library to decode tracked music files (modules) - development files
[-] openmpi-4.1.1_1            High Performance Message Passing Library
[-] openmpi-devel-4.1.1_1      High Performance Message Passing Library - development files
[-] openmpt123-0.5.10_1        Cross-platform command-line or terminal based module file player
```

It looks like the most likely candidates are `libgomp-devel-10.2.1pre1_3`, `libomp-12.0.0_3`, and `libomp-devel-12.0.0_3`... but this is the first I'd heard of `openmp`

**Comment by piratenpanda** [comment]
You have the package installed but not the devel one (if the * denotes that). I guess installing libgomp-devel as well should fix your issue

**Comment by groovenectar** [comment]
Thanks -- following this pattern, I also installed these from error messages on each attempt: `zlib-devel` `pugixml-devel` `libjpeg-turbo-devel` `libopenjpeg2-devel` `llvm` `libxml2-deve` `glib-devel`

And have landed here:

```
[dan@think build]$ cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/opt/darktable/ ..
-- Is the target platform supported: 1
-- Found little endian system. Good.
-- Building SSE2-optimized codepaths: ON
-- Performing Test C_COMPILER_UNDERSTANDS_-Wno-error=varargs
-- Performing Test C_COMPILER_UNDERSTANDS_-Wno-error=varargs - Success
-- Performing Test CXX_COMPILER_UNDERSTANDS_-Wno-error=varargs
-- Performing Test CXX_COMPILER_UNDERSTANDS_-Wno-error=varargs - Success
-- Performing Test C_COMPILER_UNDERSTANDS_-Wno-error=address-of-packed-member
-- Performing Test C_COMPILER_UNDERSTANDS_-Wno-error=address-of-packed-member - Success
-- Performing Test CXX_COMPILER_UNDERSTANDS_-Wno-error=address-of-packed-member
-- Performing Test CXX_COMPILER_UNDERSTANDS_-Wno-error=address-of-packed-member - Success
-- Looking for external programs
-- Found perl
-- Found intltool-merge
-- Found desktop-file-validate
-- Found LibXml2: /usr/lib64/libxml2.so (found version "2.9.10") 
CMake Error at /usr/lib/cmake/llvm/LLVMExports.cmake:1380 (message):
  The imported target "sancov" references the file

     "/usr/bin/sancov"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/usr/lib/cmake/llvm/LLVMExports.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /usr/lib64/cmake/llvm/LLVMConfig.cmake:278 (include)
  CMakeLists.txt:303 (find_package)
  CMakeLists.txt:311 (find_llvm)
```

**Comment by piratenpanda** [comment]
llvm provides this on arch so I guess it does the same for void

**Comment by groovenectar** [comment]
I noticed it in the source code here: https://github.com/void-linux/void-packages/blob/master/srcpkgs/llvm12/template#L231

That it's a function called `clang-tools-extra_package`

I checked for packages with `cmake` in the name and found `clang-tools-extra` which does create `/usr/bin/sancov`

Based on subsequent errors, added these packages as well: `gtk+3-devel` `sqlite-devel` `librsvg-devel` `libcurl-devel` `json-glib-devel`

Now the kicker to close out the night:

```
CMake Error at cmake/modules/FindExiv2.cmake:45 (message):
  Exiv2 version check failed.  Version was found, at least version 0.24 is
  required
Call Stack (most recent call first):
  src/CMakeLists.txt:392 (find_package)
```

```
$ exiv2 --version
exiv2 0.27.4
```

Hah!!!

**Comment by piratenpanda** [comment]
exiv2-devel ;)

**Comment by groovenectar** [comment]
Oy, thanks again! And these were the last two to cover all the errors _and_ warnings: `sdl2-devel` `libgraphicsmagick-devel`

This was anything but "quite simple" (from the condescending comment by someone else earlier).

Thanks a bunch @piratenpanda for sticking with me. I got it installed on my machine with CR3 support which is great :) :) :) but sadly I don't think this creates binaries for anyone else. That's one thing I was wondering about in https://github.com/darktable-org/darktable/issues/2170#issuecomment-917765278

An AppImage would be cool like @goeland86 mentions, though I've got no idea what's involved with packaging things like that

**Comment by piratenpanda** [comment]
Let's hope 3.8 will support cr3 files by default

**Comment by paolobenve** [comment]
what stops canon_cr3 branch from @cytrinox from beeing merged?

**Comment by svenakela** [comment]
Holy Maria's frozen nipples. I should've written down what I had to do to get CR3 support in DT, but I've lost count of all steps.
In the end I had to (among other things):
 - apt remove the installed libexiv, git clone exiv2 and compile the 0.27 branch with `cmake .. -DCMAKE_BUILD_TYPE=Release -DEXIV2_ENABLE_BMFF=On`
 - manually git switch the Cytrinox/rawspeed Canon_cr3 branch
 - do the Darktable build with `cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/opt/darktable/ -DDONT_USE_INTERNAL_LUA=Off`

It took me quite a few hours to get it running, and I am not a compile noob. Please anyone who can make a prebuilt package, please do! 

I am finally running DT with CR3 support!

**Comment by goeland86** [comment]
Very cool news, I finally managed to get exiv2 compiled with ISOBMFF in Gentoo - submitted a PR to help other Gentoo users with that.

But I'm left with a rather puzzling problem: Darktable sees Exiv2 with BMFF support, yet my EOS 90D's CR3 files only show the black skull with the message "Darktable could not load the image" and a link to check supported cameras. Yet the prebuilt binary for Windows was able to open my files just fine.

Did I miss something?

**Comment by piratenpanda** [comment]
Did you use cytrinox' rawspeed branch?

**Comment by goeland86** [comment]
Yep.

**Comment by piratenpanda** [comment]
the canon_cr3 one for sure? orherwise i cant see a reason for it not working? did you build dt master or 3.6 branch?

**Comment by goeland86** [comment]
```
~/darktable $ cat .gitmodules 
[submodule "src/external/rawspeed"]
	path = src/external/rawspeed
	url = https://github.com/cytrinox/rawspeed.git
	branch = canon_cr3
[submodule "src/external/OpenCL"]
	path = src/external/OpenCL
	url = https://github.com/KhronosGroup/OpenCL-Headers.git
[submodule "src/external/libxcf"]
	path = src/external/libxcf
	url = https://github.com/houz/libxcf.git
[submodule "src/external/whereami"]
	path = src/external/whereami
	url = https://github.com/gpakosz/whereami
[submodule "src/tests/integration"]
	path = src/tests/integration
	url = https://github.com/darktable-org/darktable-tests.git
```
Updated that right after I did the `git submodule init` so it should've pulled the right one from the get go.
Here's the output of `git status`:
```
~/darktable $ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   .gitmodules

no changes added to commit (use "git add" and/or "git commit -a")
```
I then ran `./build.sh` to generate the DT build.

**Comment by svenakela** [comment]
@goeland86 I had a build problem after getting rawspeed stuff correct and I removed the existing exiv2 (`sudo apt remove libexiv2-14`), git cloned their repo, checked out the `0.27-maintenance` branch and compiled/installed it.
I also didn't get the rawspeed to work until I actually went to the `src/external/rawspeed` folder and fetched the `canon_cr3` branch manually.

**Comment by goeland86** [comment]
Yeah I followed your recommendations on those build flags.

Let me try the manual branch checkout as you recommend.

**Comment by goeland86** [comment]
Ok, that worked. Thanks @svenakela 

Steps in order on Gentoo to get CR3 working:

- Set exiv2 with the bmff use flag active (submitted a PR for exiv2-0.27.4-r3 to include that option, should be added to repo soon I hope)
- clone darktable master
- edit `.gitmodules` to point rawspeed at cytrinox repo
- git submodule init
- git submodule update
- cd src/external/rawspeed
- git checkout canon_cr3
- cd ../../../
- ./build.sh

Darktable now supports cr3 😁 thanks to all the guides above, couldn't have made it work without everyone's contributions on this. Big thanks to @cytrinox for his work!

**Comment by groovenectar** [comment]
It would seem that Linux users are going to be having the comparatively hardest time uninstalling this thing if they ever wanted to, not to suggest that they ever should... but is there any input on potentially packaging DarkTable for Linux? Preferences?

**Comment by goeland86** [comment]
> It would seem that Linux users are going to be having the comparatively hardest time uninstalling this thing if they ever wanted to, not to suggest that they ever should... but is there any input on potentially packaging DarkTable for Linux? Preferences?

I think appimage or snap are probably the more ubiquitous ways of packaging that make it easy. Appimage can also be updated, the way freecad was doing it with appimage updater, so the user doesn't need to pull a while new image for every new version.

**Comment by groovenectar** [comment]
What do you have against Flathub? Simply that you use Snap instead?

I agree about `.AppImage`

**Comment by goeland86** [comment]
> What do you have against Flathub? Simply that you use Snap instead?

Actually I just forgot about that option 😅
Honestly I think any of the 3 mentioned is viable for pre release versions. I would go with what's easier to put together and maintain.

**Comment by piratenpanda** [comment]
May I suggest moving this discussion to https://discuss.pixls.us/

**Comment by groovenectar** [comment]
> May I suggest moving this discussion to https://discuss.pixls.us/

Sorry, is the a company behind DarkTable?

**Comment by piratenpanda** [comment]
No but thats the "official" forum for all things open source photography including darktable.

**Comment by svenakela** [comment]
Good thought. I've added more input here.
https://discuss.pixls.us/t/darktable-ubuntu-cr3-support/

**Comment by TurboGit** [comment]
> @TurboGit, we should consider locking this issue, because it turned into support forum, too much noise. There's nothing to discuss about the feature itself, just waiting for rawspeed.

Locked now.
